### PR TITLE
bpo-42513: Attempt to fix hang caused by failure to test pollfd.revents

### PR DESCRIPTION
**Pull request by BarneyStratford**
I've been stress-testing my code with a really weedy wireless dongle that drops out frequently and causes all sorts of connection issues for my code to handle gracefully. Unfortunately, the following code eventually hangs:

```
import socket
self.__socket = socket.create_connection ([host, port], 10000)
value = self.__socket.recv (4096)
```
GDB says:

```
(gdb) bt
#0  0x76d33c94 in __GI___poll (fds=0x7ea55148, nfds=1, timeout=10000)
    at ../sysdeps/unix/sysv/linux/poll.c:29
#1  0x001e8014 in poll (__timeout=<optimized out>, __nfds=<optimized out>, 
    __fds=<optimized out>, __fds=<optimized out>, __nfds=<optimized out>, 
    __timeout=<optimized out>)
    at /usr/include/arm-linux-gnueabihf/bits/poll2.h:46
#2  internal_select (writing=writing@entry=0, interval=<optimized out>, 
    connect=0, connect@entry=4156908, s=<optimized out>)
    at ../Modules/socketmodule.c:745
#3  0x001ec588 in sock_call_ex (s=s@entry=0x76979d68, writing=writing@entry=0, 
    sock_func=sock_func@entry=0x1e736c <sock_recv_impl>, data=0x7ea551b0, 
    data@entry=0x7ea551a8, connect=connect@entry=0, err=err@entry=0x0, 
    timeout=10000000000) at ../Modules/socketmodule.c:840
#4  0x001ed394 in sock_call (data=0x7ea551a8, func=0x1e736c <sock_recv_impl>, 
    writing=0, s=0x76979d68) at ../Modules/socketmodule.c:3287
#5  sock_recv_guts (s=s@entry=0x76979d68, cbuf=<optimized out>, 
    len=<optimized out>, flags=<optimized out>)
    at ../Modules/socketmodule.c:3287
#6  0x001ed51c in sock_recv (s=0x76979d68, args=<optimized out>)
    at ../Modules/socketmodule.c:3318
```

It seems that the call to poll(2) is never returning, even with a timeout specified. Googling for this problem, we get:

[https://stackoverflow.com/questions/56038224/poll-waits-indefinitely-although-timeout-is-specified](https://stackoverflow.com/questions/56038224/poll-waits-indefinitely-although-timeout-is-specified)

Looking at the source for socketmodule.c, I can see that we're not checking to see why the poll returned, and failing to handle an error event. It should be an easy fix, but because of the delicate timings, it will take several days to test both that this really is the reason for the hang and that it's the correct fix for it.

I'm creating this PR at an early stage to enable others to see what I'm doing and make comments or criticisms as required. It's not ready to merge yet.

https://bugs.python.org/issue42513

### CODE REVIEW

**Comment by jstasiak** [comment]
Is my understanding correct that the `poll()` call *is* actually returning but it's called in an infinite loop due to the error condition not being checked?

**Comment by BarneyStratford** [comment]
I've just this minute been working on this problem and yes, you're absolutely correct. It just looked like it was hanging because it was sitting in the same place with the same backtrace every time I probed this bug. I can also confirm that the hang happens exactly when revents tells us that there's a problem. (I'm running a version with lots of PySys_WriteStdouts in so I can just let it run until everything blows up and then analyse after the fact.)

File: Modules/socketmodule.c [hunk 1]
```diff
@@ -758,6 +758,8 @@ internal_select(PySocketSockObject *s, int writing, _PyTime_t interval,
 
     Py_BEGIN_ALLOW_THREADS;
     n = poll(&pollfd, 1, (int)ms);
+    if (pollfd.revents & (POLLERR | POLLNVAL))
+        n = -1;
```
Comment (benjaminp) [inline Modules/socketmodule.c:762]: The caller of `internal_select` probably expects `errno` to be set if `internal_select` returns `-1`. For `POLLINVAL`, I suppose we can just set `errno` to `EBADF`. For `POLLERR`, the best we can do may be run the outer operation and hope it returns the underlying error.

Does `POLLHUP` also need some handling here?

**Review by benjaminp — COMMENTED** [review]
This definitely needs a test.

**Comment by BarneyStratford** [comment]
I have determined that my issue was in fact caused by a documented edge-case in an area unrelated to this. As the Python code seems to be working correctly, I am cancelling this pull request. I have also closed the bug report.

**Pull request closed without merging**